guiclient/guiclient.cpp

```
@@ -1633,23 +1634,97 @@ QString translationFile(QString localestr, const QString component, QString &ver
+        if (elemThis.tagName() == filename)
+        {
+          currentVersion = elemThis.attribute("version");
+          break;
+        }
+      }
+    }
+
+    versions.close();
+  }
+
+  XSqlQuery data;
+  data.prepare("SELECT qm_version, qm_data "
+               "  FROM qm "
+               " WHERE qm_extension_name=:extension "
+               "   AND qm_lang=:lang "
+               "   AND qm_country=:country "
+               "   AND qm_version!=:version;");
+  data.bindValue(":extension", component);
+  data.bindValue(":lang", localestr.split("_")[0]);
+  data.bindValue(":country", localestr.contains("_") ? localestr.split("_")[1] : "");
+  data.bindValue(":version", currentVersion);
+  data.exec();
+  if (data.first())
   {
-    QString filename = component + "." + localestr;
-    if (DEBUG) qDebug() << "looking for translation" << dir << filename;
-    if (translator.load(filename, dir)) // this doesn't install the translation
+    QFile qm(dir + "/" + filename + ".qm");
+    if (qm.open(QIODevice::WriteOnly))
+      qm.write(data.value("qm_data").toByteArray());
+    qm.close();
+
+    if (doc.isNull())
     {
-      if (! version.isNull())
-        version = translator.translate(component.toLatin1().data(), "Version");
-      if (DEBUG)
-        qDebug() << "loadable translation" << dir << filename
-                 << "test:" << translator.translate("GUIClient", "Custom");
-      return dir + "/" + filename;
+      QDomElement root = doc.createElement("translationVersions");
+      doc.appendChild(root);
     }
+
+    if (currentVersion.isEmpty())
+    {
+      elemThis = doc.createElement(filename);
+      elemThis.setAttribute("version", data.value("qm_version").toString());
+      doc.documentElement().appendChild(elemThis);
+    }
+    else
+      elemThis.setAttribute("version", data.value("qm_version").toString());
+
+    if (versions.open(QIODevice::WriteOnly))
+      versions.write(doc.toByteArray());
```

Should we report an error if we couldn't open the file for writing or fail silently?